### PR TITLE
fix(adapter-prisma): make ?search filter work on SQLite

### PR DESCRIPTION
**Pull request by reikjarloekl**
## Summary

`PrismaStore.getFeedbacks` currently emits `where: { message: { contains, mode: "insensitive" } }` unconditionally. SQLite's Prisma connector rejects `mode: "insensitive"` and returns a 500, so any consumer running the adapter on SQLite gets a hard failure the moment a `?search=` query hits the GET endpoint.

This PR makes the `mode` field provider-aware while keeping current Postgres/MySQL/CockroachDB behaviour identical.

## Changes

- [x] `PrismaStore` accepts a new `caseInsensitiveSearch?: boolean` option (also surfaced through `createSitepingHandler({ prisma, ... })`).
- [x] When the option is omitted, the store best-effort detects the active provider via the Prisma client's internal `_activeProvider` / `_engineConfig.activeProvider` / `_engine.config.activeProvider` and only emits `mode: "insensitive"` for providers that support it (`postgresql`, `postgres`, `mysql`, `cockroachdb`).
- [x] When detection fails (unknown shape on a future Prisma release), it falls back to the historical `mode: "insensitive"` behaviour — Postgres consumers do not regress; SQLite consumers can flip the explicit option.
- [x] Six new unit tests covering: default behaviour, `caseInsensitiveSearch: false`, mocked `sqlite` / `postgresql` / `mysql` providers, and `?search` not present.
- [x] README gains a **Search and case sensitivity** subsection with the provider table and the explicit-override snippet.

No public API removed; the new option is opt-in.

## Test plan

Ran inside `packages/adapter-prisma`:

- `bun run check` — clean (whole monorepo type-checks: 10/10)
- `bun run lint` — clean
- `bun run test:run` — **1250/1250 passed**, including the 6 new `?search filter — case-insensitive mode` cases
- `bun run build` — 7/7 successful

Real-world verification: the change is the upstream version of a workaround currently shipped in a downstream consumer (Next.js + SQLite). With this fix in place, `GET /api/siteping?search=foo` returns 200 on SQLite instead of 500.

## Checklist

- [x] Tests pass (`bun run test:run`)
- [x] Types pass (`bun run check`)
- [x] No dead code or unused imports
- [x] Commit messages follow Conventional Commits (`type(scope): description`)

## Notes for reviewers

- The provider-detection helper deliberately probes three different internal paths because Prisma has moved this field across releases; the fallback path is the safe historical behaviour, so a future rename can only re-introduce the SQLite issue (it cannot regress Postgres).
- Consumers who want zero reliance on Prisma internals can pass `caseInsensitiveSearch: false` (or `true`) explicitly — detection is skipped when the option is provided.

### CODE REVIEW

File: packages/adapter-prisma/src/index.ts [hunk 1]
```diff
@@ -54,6 +54,61 @@ export interface SitepingPrismaClient {
 
 const INCLUDE_ANNOTATIONS = { annotations: true };
 
+/**
+ * Prisma datasource providers that support the `mode: "insensitive"` option
+ * on string filters. SQLite, MongoDB, and SQL Server do not — using `mode`
+ * against them raises `Unsupported('Argument mode is not supported.')` for
+ * SQLite, and similar errors elsewhere. Source: Prisma docs §case-sensitivity.
+ */
+const PROVIDERS_SUPPORTING_INSENSITIVE_MODE = new Set(["postgresql", "postgres", "mysql", "cockroachdb"]);
```
Comment (NeosiaNexus) [inline packages/adapter-prisma/src/index.ts]: **Per the codegen evidence** in the review summary: drop `mysql` (no `mode?:` in its `StringFilter`) and add `mongodb` (it does have `mode?:`).

```suggestion
const PROVIDERS_SUPPORTING_INSENSITIVE_MODE = new Set(["postgresql", "postgres", "mongodb", "cockroachdb"]);
```

The doc comment above (lines 57-62) is also worth updating — it currently says MongoDB doesn't support `mode`, which is the inverse of reality. Suggested rewrite:

```ts
/**
 * Prisma datasource providers whose generated client exposes `mode?: QueryMode`
 * on string filters. Verified against Prisma 6.x by inspecting the generated
 * `StringFilter` type per provider:
 *   - postgresql, mongodb, cockroachdb → emit `mode?: QueryMode`
 *   - mysql, sqlite, sqlserver → no `mode` field; passing it raises
 *     `PrismaClientValidationError: Unknown argument 'mode'` at runtime.
 * `postgres` is kept as a defensive alias in case `_activeProvider` ever
 * surfaces the legacy spelling.
 */
```

File: packages/adapter-prisma/src/index.ts [hunk 2]
```diff
@@ -54,6 +54,61 @@ export interface SitepingPrismaClient {
...
+    | null
+    | undefined;
+  const fromActive = candidate?._activeProvider;
+  if (typeof fromActive === "string") return fromActive;
+  const fromEngineConfig = candidate?._engineConfig?.activeProvider;
+  if (typeof fromEngineConfig === "string") return fromEngineConfig;
+  const fromEngine = candidate?._engine?.config?.activeProvider;
+  if (typeof fromEngine === "string") return fromEngine;
+  return null;
+}
```
Comment (NeosiaNexus) [inline packages/adapter-prisma/src/index.ts:98]: Wrap the probes in `try/catch` so a Proxy-based mock with throwing get-traps doesn't crash the constructor. Pre-PR the constructor was a pure assignment, so this brings it back to the same robustness level.

```suggestion
function detectActiveProvider(prisma: unknown): string | null {
  try {
    const candidate = prisma as
      | {
          _activeProvider?: unknown;
          _engineConfig?: { activeProvider?: unknown };
          _engine?: { config?: { activeProvider?: unknown } };
        }
      | null
      | undefined;
    const fromActive = candidate?._activeProvider;
    if (typeof fromActive === "string") return fromActive;
    const fromEngineConfig = candidate?._engineConfig?.activeProvider;
    if (typeof fromEngineConfig === "string") return fromEngineConfig;
    const fromEngine = candidate?._engine?.config?.activeProvider;
    if (typeof fromEngine === "string") return fromEngine;
    return null;
  } catch {
    return null;
  }
}
```

File: packages/adapter-prisma/src/index.ts [hunk 3]
```diff
@@ -62,9 +117,19 @@ const INCLUDE_ANNOTATIONS = { annotations: true };
...
-  constructor(prisma: SitepingPrismaClient) {
+  constructor(prisma: SitepingPrismaClient, options: PrismaStoreOptions = {}) {
     this.prisma = prisma;
+    if (typeof options.caseInsensitiveSearch === "boolean") {
+      this.caseInsensitiveSearch = options.caseInsensitiveSearch;
+    } else {
+      const provider = detectActiveProvider(prisma);
+      // Default true when the provider can't be detected — preserves the
+      // historical Postgres-first behaviour for unknown setups.
+      this.caseInsensitiveSearch = provider === null || PROVIDERS_SUPPORTING_INSENSITIVE_MODE.has(provider);
```
Comment (NeosiaNexus) [inline packages/adapter-prisma/src/index.ts]: Inverting the fallback default from `true` to `false` makes the worst case for an undetectable client a UX downgrade (lose non-ASCII case-insensitivity) instead of a 500 — bounded vs unbounded failure.

```suggestion
    } else {
      const provider = detectActiveProvider(prisma);
      // When the provider can't be detected, default to `false`: `contains`
      // without `mode` works on every Prisma provider; `mode: "insensitive"`
      // throws on MySQL/SQLite/SQL Server. Trades non-ASCII case-insensitivity
      // on undetectable Postgres clients (rare — _activeProvider is set on
      // every real Prisma 5/6 client) for not crashing on the others.
      this.caseInsensitiveSearch = provider !== null && PROVIDERS_SUPPORTING_INSENSITIVE_MODE.has(provider);
```

File: packages/adapter-prisma/README.md [hunk 4]
```diff
@@ -42,10 +42,37 @@ export const { GET, POST, PATCH, DELETE, OPTIONS } = createSitepingHandler({ pri
...
+### Search and case sensitivity
+
+The `?search=` filter is built with Prisma's `contains` operator. Whether it
+matches case-insensitively depends on the database provider:
+
+| Provider | Default `caseInsensitiveSearch` | Behaviour |
+|----------|---------------------------------|-----------|
+| `postgresql`, `mysql`, `cockroachdb` | `true` (auto) | Emits `mode: "insensitive"` — case-insensitive across all letters, including non-ASCII. |
+| `sqlite`, `sqlserver`, `mongodb` | `false` (auto) | No `mode` field — uses each database's default `LIKE` semantics. SQLite is ASCII-case-insensitive by default; SQL Server depends on column collation. |
+| Unknown / undetectable | `true` (auto) | Preserves historical Postgres-style behaviour. |
```
Comment (NeosiaNexus) [inline packages/adapter-prisma/README.md]: Three corrections to the matrix: MongoDB belongs in the supported bucket (Prisma codegen exposes `mode?:` for it), MySQL doesn't (it doesn't), and the unknown-fallback row should mirror the inverted constructor default from suggestion #2.

```suggestion
| `postgresql`, `mongodb`, `cockroachdb` | `true` (auto) | Emits `mode: "insensitive"` — case-insensitive across all letters, including non-ASCII. These are the providers whose generated Prisma client exposes `mode?: QueryMode` on string filters. |
| `mysql`, `sqlite`, `sqlserver` | `false` (auto) | No `mode` field (Prisma's generated client doesn't expose it for these providers — passing it raises `PrismaClientValidationError: Unknown argument 'mode'`). Falls back to each database's default `LIKE` semantics: MySQL is case-insensitive on `_ci` collations (the default); SQLite is case-insensitive on ASCII; SQL Server depends on column collation. |
| Unknown / undetectable | `false` (auto) | `contains` without `mode` works on every provider; `mode: "insensitive"` would throw on MySQL/SQLite/SQL Server. Pass `caseInsensitiveSearch: true` explicitly if you know your client is Postgres/Mongo/Cockroach but the provider auto-detection failed. |
```

**Review by NeosiaNexus — CHANGES_REQUESTED** [review]
Hey @reikjarloekl — thanks for the thorough PR! Tests, README section, edge cases all considered. The auto-detection scaffolding around `_activeProvider` / `_engineConfig` / `_engine` is exactly the right shape, and per-provider gating is the cleanest fix for this class of bug.

I dug in with empirical verification — running `prisma generate` for each provider on Prisma 6.19.3 and inspecting the generated `StringFilter` type — and there are a few corrections needed before merge. They're all tightenings on the pattern you established.

## Empirical ground truth

```
postgresql  → mode?: QueryMode  ✅  (in StringFilter)
mongodb     → mode?: QueryMode  ✅
cockroachdb → mode?: QueryMode  ✅
mysql       → no mode (emits `search?: string` instead)
sqlite      → no mode
sqlserver   → no mode
```

Reproducer (under a minute end-to-end):

```bash
mkdir -p /tmp/prisma-matrix/{pg,mysql,sqlite,mongo,sqlserver,cockroach}
# In each: minimal schema.prisma with the corresponding provider, then:
bunx prisma generate
grep -A14 'export type StringFilter<' node_modules/.prisma/client/index.d.ts
```

Also confirmed: `_activeProvider` propagates through `prisma.$extends({ name: 'test' })` — the basic extension case works fine, no need to special-case it.

## 🔴 Must-fix before merge

### 1. Provider matrix is incorrect (MySQL ↔ MongoDB swapped)

Current whitelist is `["postgresql", "postgres", "mysql", "cockroachdb"]`. Per the codegen evidence above, it should be `["postgresql", "postgres", "mongodb", "cockroachdb"]`.

Net effect of the PR as-is:
- ✅ Fixes SQLite and SQL Server (the headline)
- ❌ Doesn't fix MySQL — the whitelist still emits `mode` for it, which throws `Unknown argument 'mode'` at runtime
- ❌ Regresses MongoDB — it currently *does* receive `mode: "insensitive"` (works, gives non-ASCII case-insensitivity via `$regex`), and this PR removes that

See suggestion on line 63.

### 2. Unknown-provider fallback should default to `false`

`provider === null` currently defaults to `caseInsensitiveSearch = true`. For an undetectable client whose underlying provider is SQLite/MySQL/SQL Server (custom mock, future Prisma version where all three internal probe paths get renamed), this still hits the very 500 the PR is fixing.

Asymmetric risk argues for `false`:
- Default `true` worst case: 500 on the GET endpoint
- Default `false` worst case: undetectable Postgres client loses non-ASCII case-insensitivity

`_activeProvider` is set on every real Prisma 5/6 client today, so the practical difference is tiny — but the failure mode shifts from unbounded (crash) to bounded (UX downgrade). See suggestion on lines 127-131.

### 3. README matrix follows from #1 and #2

See suggestion on lines 56-58.

## 🟡 Polish (not blocking)

- **`try/catch` around `detectActiveProvider`.** A Proxy-based mock with throwing get-traps would crash the constructor pre-PR (which was a pure assignment). One-liner — suggestion on lines 73-89.
- **Source comment refers to the Prisma 5 error message.** Prisma 6 raises `PrismaClientValidationError: Unknown argument 'mode'. Did you mean 'lte'?` — verified empirically. Doc accuracy only; no behaviour change.
- **Commit type (purely cosmetic).** `feat(adapter-prisma):` would be a touch more accurate than `fix(adapter-prisma):` since this adds a public option, but it changes nothing version-wise: this repo sets `bump-patch-for-minor-pre-major: true` in `release-please-config.json`, so during pre-major (adapter-prisma is `0.4.4`) both `feat:` and `fix:` yield a patch bump. Pick whichever reads better to you.
- **PrismaStoreOptions docstring (lines 95-108)** lists `mysql`/`mongodb` in the wrong buckets — same root cause as #1. Easiest is to update once the matrix lands.

## 🟢 Recommended additional tests

I'd lock the matrix in by the test suite. Drop-in cases (no plumbing changes needed):

<details>
<summary>Click to expand test additions</summary>

```ts
it("auto-detects mysql via _activeProvider and omits mode", async () => {
  // MySQL's generated Prisma client does not expose `mode?:` on string
  // filters — passing it raises `Unknown argument 'mode'` at runtime.
  const mysqlPrisma = Object.assign(mockPrisma(), { _activeProvider: "mysql" });
  const mysqlHandler = createSitepingHandler({ prisma: mysqlPrisma });
  mysqlPrisma.sitepingFeedback.findMany.mockResolvedValue([]);
  mysqlPrisma.sitepingFeedback.count.mockResolvedValue(0);
  const req = new Request("http://localhost/api/siteping?projectName=test&search=hello");
  await mysqlHandler.GET(req);
  const callArgs = mysqlPrisma.sitepingFeedback.findMany.mock.calls[0][0] as {
    where: { message?: { contains: string; mode?: string } };
  };
  expect(callArgs.where.message).not.toHaveProperty("mode");
});

it("auto-detects mongodb via _activeProvider and keeps mode", async () => {
  // MongoDB's generated Prisma client exposes `mode?: QueryMode` (Prisma
  // compiles it to a case-insensitive $regex under the hood).
  const mongoPrisma = Object.assign(mockPrisma(), { _activeProvider: "mongodb" });
  const mongoHandler = createSitepingHandler({ prisma: mongoPrisma });
  mongoPrisma.sitepingFeedback.findMany.mockResolvedValue([]);
  mongoPrisma.sitepingFeedback.count.mockResolvedValue(0);
  const req = new Request("http://localhost/api/siteping?projectName=test&search=hello");
  await mongoHandler.GET(req);
  const callArgs = mongoPrisma.sitepingFeedback.findMany.mock.calls[0][0] as {
    where: { message?: { contains: string; mode?: string } };
  };
  expect(callArgs.where.message).toEqual({ contains: "hello", mode: "insensitive" });
});

it("auto-detects cockroachdb via _activeProvider and keeps mode", async () => {
  const cockroachPrisma = Object.assign(mockPrisma(), { _activeProvider: "cockroachdb" });
  const cockroachHandler = createSitepingHandler({ prisma: cockroachPrisma });
  cockroachPrisma.sitepingFeedback.findMany.mockResolvedValue([]);
  cockroachPrisma.sitepingFeedback.count.mockResolvedValue(0);
  const req = new Request("http://localhost/api/siteping?projectName=test&search=hello");
  await cockroachHandler.GET(req);
  const callArgs = cockroachPrisma.sitepingFeedback.findMany.mock.calls[0][0] as {
    where: { message?: { contains: string; mode?: string } };
  };
  expect(callArgs.where.message).toEqual({ contains: "hello", mode: "insensitive" });
});

it("auto-detects sqlserver via _activeProvider and omits mode", async () => {
  const mssqlPrisma = Object.assign(mockPrisma(), { _activeProvider: "sqlserver" });
  const mssqlHandler = createSitepingHandler({ prisma: mssqlPrisma });
  mssqlPrisma.sitepingFeedback.findMany.mockResolvedValue([]);
  mssqlPrisma.sitepingFeedback.count.mockResolvedValue(0);
  const req = new Request("http://localhost/api/siteping?projectName=test&search=hello");
  await mssqlHandler.GET(req);
  const callArgs = mssqlPrisma.sitepingFeedback.findMany.mock.calls[0][0] as {
    where: { message?: { contains: string; mode?: string } };
  };
  expect(callArgs.where.message).not.toHaveProperty("mode");
});

it("reads provider from _engineConfig.activeProvider when _activeProvider is missing", async () => {
  const altPrisma = Object.assign(mockPrisma(), { _engineConfig: { activeProvider: "postgresql" } });
  const altHandler = createSitepingHandler({ prisma: altPrisma });
  altPrisma.sitepingFeedback.findMany.mockResolvedValue([]);
  altPrisma.sitepingFeedback.count.mockResolvedValue(0);
  const req = new Request("http://localhost/api/siteping?projectName=test&search=hello");
  await altHandler.GET(req);
  const callArgs = altPrisma.sitepingFeedback.findMany.mock.calls[0][0] as {
    where: { message?: { contains: string; mode?: string } };
  };
  expect(callArgs.where.message).toEqual({ contains: "hello", mode: "insensitive" });
});

it("survives a prisma client whose provider probe throws", async () => {
  // Some test doubles use Proxies whose get-traps throw — the constructor
  // must still build a working store rather than crash.
  const throwingPrisma = new Proxy(mockPrisma(), {
    get(target, prop, receiver) {
      if (prop === "_activeProvider" || prop === "_engineConfig" || prop === "_engine") {
        throw new Error("probe denied");
      }
      return Reflect.get(target, prop, receiver);
    },
  }) as unknown as ReturnType<typeof mockPrisma>;
  const throwingHandler = createSitepingHandler({ prisma: throwingPrisma });
  throwingPrisma.sitepingFeedback.findMany.mockResolvedValue([]);
  throwingPrisma.sitepingFeedback.count.mockResolvedValue(0);
  const req = new Request("http://localhost/api/siteping?projectName=test&search=hello");
  await throwingHandler.GET(req);
  const callArgs = throwingPrisma.sitepingFeedback.findMany.mock.calls[0][0] as {
    where: { message?: { contains: string; mode?: string } };
  };
  // Probe threw → fallback → no mode (safe default).
  expect(callArgs.where.message).not.toHaveProperty("mode");
});
```

Heads-up: the existing **"emits mode:'insensitive' for the default Postgres-style client"** test relies on the fallback path (the mock has no `_activeProvider`). With suggestion #2 applied, that test's expectation flips — easiest is to rename it to **"falls back to omitting mode when the active provider can't be detected"** and assert `not.toHaveProperty("mode")`.

</details>

## What I verified locally

I checked out the branch and applied all four suggestions to confirm everything stays green:

- `bun run check` — 10/10 packages ✅
- `bun run lint` (biome) — 0 issues ✅
- `bun run test:run` — 1256 tests passed (1250 + 6 new) ✅
- Real Prisma + SQLite end-to-end test — auto-detects sqlite, no `mode` emitted, `?search=hello` returns the expected row, control case (forced `caseInsensitiveSearch: true` on SQLite) still throws ✅

---

Thanks again — this is a real fix, and the bones are right. Once the matrix is corrected, this is a clean ship. Happy to take another pass once you push.

**Comment by reikjarloekl** [comment]
Thanks for the thorough review — the empirical check (running `prisma generate` per provider and inspecting the generated `StringFilter`) was the deciding evidence, and you were right on every point.

Pushed `eeccc15` with all four asks applied:

- **Provider matrix corrected** — `mysql` removed, `mongodb` added. Set is now `{postgresql, postgres, mongodb, cockroachdb}`. The doc comment, `PrismaStoreOptions` JSDoc, and README matrix table all match.
- **Unknown-provider fallback inverted** from `true` → `false`. Reasoning matches yours: `contains` without `mode` works on every provider, while `mode: "insensitive"` throws on MySQL/SQLite/SQL Server — bounded vs unbounded failure mode.
- **`try/catch` around `detectActiveProvider`** so a Proxy-based mock with throwing get-traps doesn't crash the constructor (restores pre-PR robustness).
- **6 new tests** locking the matrix into CI: `mysql` / `mongodb` / `cockroachdb` / `sqlserver` auto-detect, `_engineConfig.activeProvider` fallback path, and throwing-Proxy survival. The existing Postgres-default test was renamed to assert the inverted unknown-fallback behaviour.

Local verification: `bun run check` 10/10, `biome check packages/adapter-prisma` clean, `bun run test:run` 1256/1256.

Ready for another look whenever you have a moment.

**Comment by NeosiaNexus** [comment]
Took the rebase off your plate, @reikjarloekl — main moved underneath you while we were iterating (#55 added page-scoped `url`/`urlPattern` filters, #58 added pluggable screenshot storage, both touched the same lines you did). The branch is now rebased on `main` with both of your commits preserved as-authored:

- `33b407a` fix(adapter-prisma): make ?search filter work on SQLite
- `7723f06` fix(adapter-prisma): correct provider matrix per reviewer feedback

Conflict resolution (three spots, all mechanical):

- `PrismaStoreOptions` now carries both `caseInsensitiveSearch` and `screenshotStorage` so consumers can configure either independently
- `getFeedbacks` `where` builder keeps the new `url` / `urlPattern` filters alongside your conditional `mode` logic
- README query-param table keeps the new `url` / `urlPattern` rows alongside your "Search and case sensitivity" section

Local verification on the rebased branch:

- `bun run test:run` — **1306/1306 passed** (your 6 new cases all green)
- `bun run check` — 10/10 packages
- `bun run lint` — clean

Thanks again for the rigour on the follow-up — running `prisma generate` per provider to verify the actual `StringFilter` shape was exactly the right move, and the `try/catch` around `detectActiveProvider` plus the throwing-Proxy test are the kind of robustness I love seeing in a contribution. Hope you'll send more.

Ready to merge from my side.

**Review by NeosiaNexus — APPROVED** [review]
Approving — all four review items applied (provider matrix corrected to {postgresql, postgres, mongodb, cockroachdb}, fallback inverted to false, try/catch around the probe, six new tests), and the rebase on `main` now has 1306/1306 passing locally. Cleared to merge once CI is green.
